Yda-6312 adding local json validation and upgrading it to V4

FuHsinyu opened the pull request:
What work has been done:
1. Changed schema validation from against remote server to local validation
2. Upgraded V3 schema to V4 schema
3. Tested in various env
4. For a few keys in JSON that are clearly not in a right order (alphabetically), I sorted them 

What tests have been done:
1. Tested successfully with all-in-one setup to local VM server (Ubuntu 24). This ensure the provider JSON schema is validated correctly.
2. Tested successfully  with Full set up to local VM server (Ubuntu 24). This ensure the Consumer-side JSON schema is also validated correctly. However, a known bug is https://utrechtuniversity.atlassian.net/browse/YDA-6327 (workaround is included in the ticket).
3.  Tested successfully with docker set up by rebuilding the ICAT images ~~(while other images are pulled from the cloud). This ensures the Docker related Schema validation is correct. Note, if you try to rebuild the Davords docker image locally, you may encounter an error. (I got it but did not create a ticket yet, since I have tried to reproduce it elsewhere. If you do encounter the same error, we can create a ticket for that then)~~ **This bug has been fixed, and a docker compose with full rebuilt images is successful**.  The complete docker rebuilt can also be tested in CI when trying to merge to the dev branch. 
5. Test against the UTest server was not done, since it requires the dev branch include the updated schemas. 

Tips for testing:
1. You need a Fresh VM to do the tests. I try to overwrite the V3 schema of the current VM but it creates more issues. So I discussed with Lazlo and we'll need a fresh VM to deploy it for development purpose. 

Other notes:
1. There are a few key entries that were updated with a new name or discarded. I have also double checked its origin. The log:  in server_config, one key is deprecated in v4 'xmsg_port' (which was an v3 only schema). Also 'host_resolution', 'host_access_control' renaming the key names as in v4 the key names have changed

For the official V4 schema template, refer to https://github.com/irods/irods/tree/4.3.4/schemas/configuration/v4
For the official V3 schema template, refer to https://github.com/irods/irods/tree/4-2-stable/configuration_schemas/v3

2. Fyi, I found a IRODS scrip that is used to upgrade schema, but that was not explained anywhere (or mayby i didn't find it). If you are interested, see https://github.com/irods/irods/blob/4.3.4/scripts/irods/upgrade_configuration.py **I also cross validated the schema with what the script intends to do. That's why some keys are deprecated, they can be traced back to the script.**

FuHsinyu committed:
update schema version from 3 to 4 and add local json validation

FuHsinyu committed:
Merge branch 'development' into yda-6312-bug-fix-for-adding-local-json-validation

FuHsinyu committed:
add prefix for irods_shcema_uri

FuHsinyu committed:
update server and irods_env config

FuHsinyu committed:
fix missing comma

FuHsinyu committed:
convert string to interger for certain key

FuHsinyu committed:
Merge branch 'development' into yda-6312-bug-fix-for-adding-local-json-validation

FuHsinyu committed:
Test docket build in git

FuHsinyu committed:
Ensure docker build in git with everytime push

FuHsinyu committed:
add docker schema validation

FuHsinyu committed:
add consumer side schema validation

FuHsinyu committed:
move host_access_control to server_config level

FuHsinyu committed:
improve consumer side json schema validation

FuHsinyu committed:
upgrade service_account_environment.json

FuHsinyu committed:
update key names created in v3 to v4

FuHsinyu committed:
update keys for consumer side json

FuHsinyu committed:
sort out irods_env and server_config (Docker) alphabetically

FuHsinyu committed:
add back removed irods_host etc

FuHsinyu committed:
remove unused comments

FuHsinyu commented on docker/images/yoda_irods_icat/server_config.json:
```diff
@@ -102,8 +110,8 @@
     "schema_name": "server_config",
-    "schema_validation_base_uri": "off",
-    "schema_version": "v3",
+    "schema_validation_base_uri": "file:///var/lib/irods/configuration_schemas/",
```
The reason I add 'file:/' is that Uri without prefix 'file:/' caused error in the created JSON that schema validation will fail.

FuHsinyu committed:
Remove upgrade_schema_config/ from remote history

stsnel commented on docker/images/yoda_irods_icat/server_config.json:
```diff
@@ -87,13 +95,13 @@
                 "shared_memory_instance": "irods_rule_language_rule_engine"
             },
             {
-                "instance_name": "irods_rule_engine_plugin-python-instance",
```
The rule engine list controls in which order the rule engine plugins will be consulted. We generally control policies using iRODS rule language and Python, so the C++ default rule engine plugin priority should not be increased relative to the Python one.

stsnel commented on roles/irods_icat/templates/setup_irods_provider.json.j2:
```diff
@@ -46,14 +30,38 @@
+      "network": "info",
+      "resource": "info",
+      "rule_engine": "info",
+      "server": "info",
```
We need debug level for the server in order to print log output from rule code.

stsnel commented on roles/irods_resource/templates/setup_irods_consumer.json.j2:
```diff
@@ -46,21 +31,46 @@
+      "network": "info",
+      "resource": "info",
+      "rule_engine": "info",
+      "server": "info",
```
We need debug level for the server in order to print log output from the rules.

stsnel reviewed (commented):
This looks good in general, I think :+1: 

When I tested the Docker setup, the provider container failed on startup with client negotiation errors. Can you have a look if you can reproduce this issue?

I've also added a few specific comments above.

FuHsinyu committed:
change server log level to debug, and adjust order of keys

FuHsinyu commented on roles/irods_resource/templates/setup_irods_consumer.json.j2:
```diff
@@ -93,21 +103,45 @@
   },
-  "default_resource_name": "{{ irods_default_resc }}",
-  "default_resource_directory": ""
+  "service_account_environment": {
```
Perhaps I should not move this key to the end of the json, since it is code refactoring?

FuHsinyu committed:
fix client negotiation error in docker

FuHsinyu committed:
rename VERSION.json to version.json

FuHsinyu committed:
rename hosts_config to host_resolution, host_access_control_config to host_access_control

FuHsinyu commented on docker/images/yoda_irods_icat/irods-icat-init.sh:
```diff
@@ -84,17 +84,19 @@ export PGPASSWORD=yodadev
-INSTALL_TIMESTAMP=$(date +'%Y-%m-%dT%H:%M:%S.000000')
-cat > /var/lib/irods/VERSION.json << VERSION
+INSTALL_TIMESTAMP=$(date -u +'%Y-%m-%dT%H:%M:%S.000000Z')  
+cat > /var/lib/irods/version.json << VERSION
```
Explaination: VERSION.json is no longer used, instead, version.json is used.

FuHsinyu committed:
Rename key names based on irods's upgrade_config script

FuHsinyu commented:
> This looks good in general, I think 👍
> 
> When I tested the Docker setup, the provider container failed on startup with client negotiation errors. Can you have a look if you can reproduce this issue?
> 
> I've also added a few specific comments above.

The error has been resolved. See bold text in the description.

stsnel committed:
Remove outdated config files from Docker setup

These configuration files are no longer used in iRODS 4.3.4. iRODS uses
the server_config.json file for these config values now.

stsnel reviewed (approved):
Excellent, looks good to me! :+1:
